app/com/linkedin/drelephant/ElephantRunner.java

```diff
@@ -188,10 +310,160 @@ private void waitInterval() {
     }
   }
 
+  private void stopFurtherAnalysis(String appId) {
+    if (_analyticJobScheduledFutureMap.containsKey(appId)) {
+      _analyticJobScheduledFutureMap.remove(appId).cancel(false);
+    }
+  }
+
+  public int getCompletedThreadPoolQueueSize() {
+    return _completedJobThreadPoolExecutor.getQueue().size();
+  }
+
+  public int getUndefinedThreadPoolQueueSize() {
+    return _undefinedJobThreadPoolExecutor.getQueue().size();
+  }
+
+  private void saveOrUpdate(AppResult result) {
+    AppResult jobInDb = AppResult.find.byId(result.id);
+    if (jobInDb == null) {
+      result.save();
+    } else {
+      // Delete and save. Prevents Optimistic Locking
+      jobInDb.delete();
+      result.save();
+    }
+  }
+
+  /**
+   * Kill the Dr. Elephant daemon
+   */
   public void kill() {
     _running.set(false);
-    if (_service != null) {
-      _service.shutdownNow();
+    if (_completedJobThreadPoolExecutor != null) {
+      _completedJobThreadPoolExecutor.shutdownNow();
+    }
+    if (_undefinedJobThreadPoolExecutor != null) {
+      _undefinedJobThreadPoolExecutor.shutdownNow();
+    }
+  }
+
+  public boolean isKilled() {
+    if (_completedJobThreadPoolExecutor != null && _undefinedJobThreadPoolExecutor != null) {
+      return !_running.get() && _completedJobThreadPoolExecutor.isShutdown() &&
+          _undefinedJobThreadPoolExecutor.isShutdown();
+    }
+    return true;
+  }
+
+  /**
+   * Analyze Completed jobs
+   */
+  private class CompletedExecutorJob implements Runnable {
+
+    private AnalyticJob _analyticJob;
+
+    CompletedExecutorJob(AnalyticJob analyticJob) {
+      this._analyticJob = analyticJob;
+      _analyticJob.setAnalysisStartTime(System.currentTimeMillis());
+    }
+
+    @Override
+    public void run() {
+      try {
+        logger.info(String.format("Analyzing %s %s", _analyticJob.getAppType().getName(), _analyticJob.getAppId()));
+        String finalStatus = _analyticJob.getFinalStatus();
+        if (finalStatus.equals(SUCCEEDED) || finalStatus.equals(FAILED)) {
+          // The finalStatus obtained from the Resource Manager Rest api is inconsistent with the final
+          // jobStatus as obtained from the mapreduce client api. Making them same will make it easier
+          // for us to handle the different cases(completed and running).
+          _analyticJob.setJobStatus(finalStatus);
+        }
+        String appState = _analyticJob.getJobStatus();
+        String analysisName =
+            String.format("%s %s %s", _analyticJob.getAppType().getName(), _analyticJob.getAppId(), appState);
+        logger.info(String.format("%s %s %s", _analyticJob.getAppType().getName(), _analyticJob.getAppId(), appState));
+
+        if (appState.equals(SUCCEEDED) || appState.equals(FAILED) || appState.equals(KILLED)) {
+          AppResult result = _analyticJob.getAnalysis(null);
+          stopFurtherAnalysis(_analyticJob.getAppId());
+          saveOrUpdate(result);
```

Comment:
You need to ensure the current scheduled job in _undefinedThreadPoolExecutor for this appId is finished before saving the result from the CompletedJob. Otherwise, it might overwrite it. stopFurtherAnalysis doesn't seem to ensure that.